### DISCUSSION TITLE
Starter Cluster with LetsEncrypt on AWS using Terraform fails to initialize and start

### DISCUSSION THREAD

**Discussion by sklirg** [gravitational/teleport#10388] [Q&A] [answered]
Hi there.

I have followed https://goteleport.com/docs/setup/deployments/aws-terraform/ and the reference terraform config here: https://github.com/gravitational/teleport/tree/master/examples/aws/terraform/starter-cluster

I realise that the referenced configuration is a HA deployment, and the reference configuration I used is a single-node deployment, but I still find the encountered errors a bit weird.

Using AMI: "ami-0ef645c60c9e0fb73"

When the cluster starts, issuing `journalctl -u teleport` results in error messages saying:

> ...
> teleport-all-pre-start[8553]:  ++ USE_LETSENCRYPT=true
> teleport-all-pre-start[8553]:  ++ USE_ACM=false
> teleport-all-pre-start[8553]:  + [[ true != \t\r\u\e ]]
> teleport-all-pre-start[8553]:  + /bin/aws s3 sync s3://<S3_BUCKET_NAME>/live/<DOMAIN> /var/lib/teleport
> systemd[1]: Started Teleport Service.
> teleport[3055]: ERROR: https private key does not exist: /var/lib/teleport/privkey.pem

So it seems to me that Teleport tries to fetch some certificate from s3, even though there are no certificates there, and it is told to use LetsEncrypt for certificates.

When trying to debug this issue, I noticed that the configuration file contains the following excerpt:

```yaml
proxy_service:
  ..
  https_keypairs:
    - /var/lib/teleport/fullchain.pem
    - /var/lib/teleport/privkey.pem
```

and **no** `acme` key.

I figured that if I remove the `https_keypairs` option and add the `acme` option with `enabled: yes` and `email: <myemail>`, it should work.

At least the logs now showed that Teleport started, but it was unable to get a certificate. This is likely because LetsEncrypt TLS challenges only work on port 443, and the Teleport configuration file specifies Teleport proxy to listen on port 3080, and the AWS terraform configuration deploys a Security Group which does **not** allow communication on port 443. So, updating the Teleport configuration to use port 443 and allow it in the security group made Teleport able to fetch a certificate through LetsEncrypt and use it.

So now Teleport started and allowed me to sign up.

However, the certificate did not end up in the S3 bucket.

The following might be related to the difference between Starter Cluster and HA Cluster, but I'll bring it up anyways:

In the Use ACM section of the docs, it says:

> use_acm
> Setting export TF_VAR_use_acm="false"
> 
> If set to the string "false", Terraform will use [LetsEncrypt](https://letsencrypt.org/) to provision the public-facing web UI certificate for the Teleport cluster ([route53_domain](https://goteleport.com/docs/setup/deployments/aws-terraform/#route53_domain) - so https://teleport.example.com/ in this example). This uses an [AWS network load balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) to load-balance connections to the Teleport cluster's web UI, and its SSL termination is handled by Teleport itself.

However, I cannot seem to find any terraform configuration to deploy a Network Load Balancer in the Starter Cluster reference. And since the Starter Cluster is configured to run on port 3080, a LetsEncrypt Challenge won't work to distribute certificates since LetsEncrypt TLS challenge only runs on port 443.
Finally, as mentioned above, if I configure Teleport to use acme "manually" (by editing the config file and adding the acme section), the certificate doesn't end up in S3 for later use.

I would appreciate some pointers here, if it's simply me misunderstanding something or if something is missing in the starter reference! :)

**Comment by scottpgallagher** [comment]
Network load balancers should be found in this terraform example code in Teleport github: 
https://github.com/gravitational/teleport/blob/740d184d1cfc69ae2e96c50ee738b13884fb232b/examples/aws/terraform/ha-autoscale-cluster/proxy_network.tf#L131

**Comment by webvictim** [comment, marked as answer]
The starter cluster is designed to deploy all Teleport services on a single instance and as such, it lacks a lot of the more tested HA functionality (which assumes that load balancers will be configured on port 443, certs will end up in the S3 bucket, etc). You've essentially found an untested gap in our examples - my apologies.

It seems like you've fixed most of your issues through experimentation which is great.

**Comment by sklirg** [comment]
Hi again!

Thanks for the replies, I continued working with this on-and-off for the past week so I'm slowly getting to a working state. 🙂 I was planning on contributing my final findings after I got it fully working, but expected it to take a bit less time. 😉 

This setup isn't going to be used for many servers or users, so going full-blown HA seems overkill to me. I'm also going to replicate the setup for a testing and production environment, so running two HA clusters seems even more overkill. 😄

I went over to using ACM in a one-node deployment. However, it needs two load balancers, one ALB for the web application GUI and an NLB for the service proxying. 

I'll try to see if I can come back and answer some of my questions when I've figured everything out. :)

**Reply by webvictim** [reply]
One thing that's worth noting is that AWS now allows you to deploy TLS listeners with ACM certs on NLBs (which it didn't at the time the Terraform was written)

As such, you should be able to just have one NLB with a TLS listener for port 443 and TCP listeners for all other ports. This would remove the need for an ALB and make the setup a little simpler.

**Reply by sklirg** [reply]
Thanks for the tip! That worked nicely. :)

**Comment by sklirg** [comment]
Hi again.

I've changed most of the setup since the initial post, and I'm currently running Teleport with ACM (AWS Cert Manager) and a single Load Balancer with listener/target groups on the required ports (https w/ACM cert@443, tcp@3024-3026).

We're running with database and kubernetes access, and this seems to work fine. I'm not sure about the Let's Encrypt issues, but I might have a stab at that again later.

Thanks again for the help 👍
